### PR TITLE
dmd.dtoh: export genCppHdrFiles to C++

### PR DESCRIPTION
**Pull request by ibuclaw**
Follow-up to #10801.

This allows `genCppHdrFiles` to be called from gdc.

### CODE REVIEW

File: src/dmd/dtoh.d [hunk 1]
```diff
@@ -193,14 +188,14 @@ private void initialize()
...
 }
 
-void genCppHdrFiles(ref Modules ms)
+extern(C++) void genCppHdrFiles(ref Modules ms)
 {
     initialize();
 
     OutBuffer buf;
-    buf.writestring("#pragma once\n");
+    buf.printf("// Automatically generated by %s Compiler v%d\n", __VENDOR__.ptr, __VERSION__);
```
Comment (thewilsonator) [inline src/dmd/dtoh.d]: Wont this use the info of the compiler used to compile the compiler?

**Comment by MoonlightSentinel** [comment]
Most of these changes are irrelevant to the C++ interface...

Also why move the ‚pragma once‘ After the comment?

File: src/dmd/dtoh.d (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/dtoh.d]: I forget that dmd doesn't bootstrap itself by default.  It's likely far too late for me to think anymore, and now you mention it, using internal global params directly makes more sense.

**Comment by ibuclaw** [comment]
> Most of these changes are irrelevant to the C++ interface...

Rationale of these three commits are:

1. FTBFS: dmd/mars.d doesn't exist in gdc source tree
2. FTBFS: Couldn't call genCppHdrFiles from C++
3. Allow for gdc to advertise itself as the generator.

> Also why move the ‚pragma once‘ After the comment?

Usually (least in every other tool I've used and/or written) the generated file notice is on the first line.  I found it surprising that it wasn't.

**Comment by MoonlightSentinel** [comment]
> Rationale of these three commits are:

Thanks for the explanation. These changes seem reasonable but the PR title is a bit misleading.

> Usually (least in every other tool I've used and/or written) the generated file notice is on the first line. I found it surprising that it wasn't.

Okay, improving user experience is always nice.

**Review by wilzbach — APPROVED** [review]

**Pull request merged**